File: server/src/test/java/dev/datarun/server/projection/ProjectionEquivalenceTest.java
package dev.datarun.server.projection;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.datarun.server.AbstractIntegrationTest;
import dev.datarun.server.identity.AliasCache;
import dev.datarun.server.subject.SubjectProjection;
import dev.datarun.server.subject.SubjectSummary;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;

import java.io.InputStream;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Projection equivalence test (E7): server PE must produce byte-identical
 * canonical output to Dart PE given the same ordered event set.
 *
 * Loads shared fixture from contracts/fixtures/projection-equivalence.json.
 * Both Java and Dart tests compare to the same expected_output.
 */
class ProjectionEquivalenceTest extends AbstractIntegrationTest {

    @Autowired
    private JdbcTemplate jdbc;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private SubjectProjection subjectProjection;

    @Autowired
    private AliasCache aliasCache;

    @BeforeEach
    void cleanDb() {
        jdbc.execute("DELETE FROM events");
        jdbc.execute("DELETE FROM subject_aliases");
        jdbc.execute("DELETE FROM subject_lifecycle");
        jdbc.execute("ALTER SEQUENCE events_sync_watermark_seq RESTART WITH 1");
    }

    @Test
    void projectionEquivalence_sameFixture_producesExpectedOutput() throws Exception {
        // Load the shared fixture
        JsonNode fixture;
        try (InputStream is = getClass().getResourceAsStream("/fixtures/projection-equivalence.json")) {
            assertThat(is).as("Fixture file must exist on classpath").isNotNull();
            fixture = objectMapper.readTree(is);
        }

        // Insert events in order (with pre-assigned sync_watermarks)
        JsonNode events = fixture.get("events");
        for (JsonNode event : events) {
            insertEvent(event);
        }

        // Insert aliases
        JsonNode aliases = fixture.get("aliases");
        for (JsonNode alias : aliases) {
            insertAlias(alias);
        }

        // Refresh alias cache so PE picks them up
        aliasCache.refresh();

        // Run server PE
        List<SubjectSummary> subjects = subjectProjection.listSubjects();

        // Extract canonical output
        JsonNode expectedOutput = fixture.get("expected_output");
        JsonNode expectedSubjects = expectedOutput.get("subjects");

        assertThat(subjects).hasSize(expectedSubjects.size());

        for (int i = 0; i < expectedSubjects.size(); i++) {
            JsonNode expected = expectedSubjects.get(i);
            SubjectSummary actual = subjects.get(i);

            assertThat(actual.id())
                    .as("subject[%d].subject_id", i)
                    .isEqualTo(expected.get("subject_id").asText());

            assertThat(actual.eventCount())
                    .as("subject[%d].event_count", i)
                    .isEqualTo(expected.get("event_count").asInt());

            assertThat(actual.flagCount())
                    .as("subject[%d].flag_count", i)
                    .isEqualTo(expected.get("flag_count").asInt());

            // Compare timestamps — both should represent the same instant
            String expectedTs = expected.get("latest_timestamp").asText();
            OffsetDateTime expectedDt = OffsetDateTime.parse(expectedTs,
                    DateTimeFormatter.ISO_DATE_TIME.withZone(ZoneOffset.UTC));
            assertThat(actual.latestTimestamp().toInstant())
                    .as("subject[%d].latest_timestamp", i)
                    .isEqualTo(expectedDt.toInstant());
        }
    }

    private void insertEvent(JsonNode event) {
        UUID id = UUID.fromString(event.get("id").asText());
        String type = event.get("type").asText();
        String shapeRef = event.get("shape_ref").asText();
        String activityRef = event.has("activity_ref") && !event.get("activity_ref").isNull()
                ? event.get("activity_ref").asText() : null;
        JsonNode subjectRef = event.get("subject_ref");
        JsonNode actorRef = event.get("actor_ref");
        UUID deviceId = UUID.fromString(event.get("device_id").asText());
        int deviceSeq = event.get("device_seq").asInt();
        long syncWatermark = event.get("sync_watermark").asLong();
        OffsetDateTime timestamp = OffsetDateTime.parse(event.get("timestamp").asText(),
                DateTimeFormatter.ISO_DATE_TIME.withZone(ZoneOffset.UTC));
        JsonNode payload = event.get("payload");

        // Insert directly with pre-assigned sync_watermark (bypass sequence)
        jdbc.update("""
                INSERT INTO events (id, type, shape_ref, activity_ref, subject_ref, actor_ref,
                                    device_id, device_seq, sync_watermark, timestamp, payload)
                VALUES (?::uuid, ?, ?, ?, ?::jsonb, ?::jsonb, ?::uuid, ?, ?, ?::timestamptz, ?::jsonb)
                """,
                id.toString(), type, shapeRef, activityRef,
                subjectRef.toString(), actorRef.toString(),
                deviceId.toString(), deviceSeq, syncWatermark,
                timestamp.toString(), payload.toString());
    }

    private void insertAlias(JsonNode alias) {
        String retiredId = alias.get("retired_id").asText();
        String survivingId = alias.get("surviving_id").asText();
        jdbc.update("""
                INSERT INTO subject_aliases (retired_id, surviving_id, merged_at)
                VALUES (?::uuid, ?::uuid, NOW())
                ON CONFLICT (retired_id) DO NOTHING
                """, retiredId, survivingId);
    }
}

File: mobile/test/projection_equivalence_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:datarun_mobile/data/event_store.dart';
import 'package:datarun_mobile/data/projection_engine.dart';
import 'package:datarun_mobile/domain/event.dart';

/// Projection equivalence test (E7): Dart PE must produce identical
/// canonical output to Java PE given the same ordered event set.
///
/// Loads shared fixture from contracts/fixtures/projection-equivalence.json.
void main() {
  sqfliteFfiInit();
  databaseFactory = databaseFactoryFfi;

  late EventStore store;
  late ProjectionEngine pe;
  late String dbPath;

  setUp(() async {
    dbPath =
        '${Directory.systemTemp.path}/datarun_equiv_${DateTime.now().microsecondsSinceEpoch}.db';
    store = EventStore(dbPath: dbPath);
    pe = ProjectionEngine(store);
  });

  tearDown(() async {
    await store.close();
    try {
      File(dbPath).deleteSync();
    } catch (_) {}
  });

  test('E7: projection equivalence — Dart PE matches shared fixture', () async {
    // Load the shared fixture
    final fixtureFile = File('${Directory.current.path}/../contracts/fixtures/projection-equivalence.json');
    expect(fixtureFile.existsSync(), isTrue,
        reason: 'Fixture file must exist at contracts/fixtures/projection-equivalence.json');
    final fixture = jsonDecode(fixtureFile.readAsStringSync()) as Map<String, dynamic>;

    // Insert events
    final events = fixture['events'] as List<dynamic>;
    for (final eventJson in events) {
      final e = eventJson as Map<String, dynamic>;
      final event = Event(
        id: e['id'] as String,
        type: e['type'] as String,
        shapeRef: e['shape_ref'] as String,
        activityRef: e['activity_ref'] as String?,
        subjectRef: Map<String, String>.from(e['subject_ref'] as Map),
        actorRef: Map<String, String>.from(e['actor_ref'] as Map),
        deviceId: e['device_id'] as String,
        deviceSeq: e['device_seq'] as int,
        syncWatermark: e['sync_watermark'] as int?,
        timestamp: e['timestamp'] as String,
        payload: Map<String, dynamic>.from(e['payload'] as Map),
      );
      await store.insert(event);
    }

    // Insert aliases
    final aliases = fixture['aliases'] as List<dynamic>;
    for (final alias in aliases) {
      final a = alias as Map<String, dynamic>;
      await store.upsertAlias(
        a['retired_id'] as String,
        a['surviving_id'] as String,
        DateTime.now().toUtc().toIso8601String(),
      );
    }

    // Run Dart PE
    final subjects = await pe.getSubjectList();

    // Compare to expected output
    final expectedSubjects =
        (fixture['expected_output']['subjects'] as List<dynamic>)
            .cast<Map<String, dynamic>>();

    expect(subjects.length, equals(expectedSubjects.length),
        reason: 'Subject count mismatch');

    for (var i = 0; i < expectedSubjects.length; i++) {
      final expected = expectedSubjects[i];
      final actual = subjects[i];

      expect(actual.subjectId, equals(expected['subject_id']),
          reason: 'subject[$i].subject_id');
      expect(actual.captureCount, equals(expected['event_count']),
          reason: 'subject[$i].event_count');
      expect(actual.flagCount, equals(expected['flag_count']),
          reason: 'subject[$i].flag_count');

      // Compare timestamps as instants
      final expectedTs =
          DateTime.parse(expected['latest_timestamp'] as String);
      final actualTs = DateTime.parse(actual.latestTimestamp);
      expect(actualTs.toUtc(), equals(expectedTs.toUtc()),
          reason: 'subject[$i].latest_timestamp');
    }
  });
}
